muikku-core-plugins/src/main/resources/META-INF/resources/scripts/src/actions/workspaces/index.ts

```diff
@@ -1383,21 +1678,91 @@ let deleteCurrentWorkspaceImage:DeleteCurrentWorkspaceImageTriggerType = functio
         }
       });
     } catch (err) {
-      if (!(err instanceof MApiError)){
+      if (!(err instanceof MApiError)) {
         throw err;
       }
-      dispatch(displayNotification(getState().i18n.text.get('TODO ERRORMSG failed to delete workspace image'), 'error'));
+      dispatch(displayNotification(getState().i18n.text.get('plugin.workspace.notification.workspaceImage.delete.error'), 'error'));
     }
   }
 }
 
-let copyCurrentWorkspace:CopyCurrentWorkspaceTriggerType = function copyCurrentWorkspace(data){
-  return async (dispatch:(arg:AnyActionType)=>any, getState:()=>StateType)=>{
+let createWorkspace: CreateWorkspaceTriggerType = function createWorkspace(data) {
+  return async (dispatch: (arg: AnyActionType) => any, getState: () => StateType) => {
     try {
-      let state:StateType = getState();
-      let currentWorkspace:WorkspaceType = getState().workspaces.currentWorkspace;
-      let cloneWorkspace:WorkspaceType = <WorkspaceType>(await promisify(mApi().workspace.workspaces
-          .create(
+      let workspace: WorkspaceType = <WorkspaceType>(await promisify(mApi().workspace.workspaces
+        .create(
+          {
+            name: data.name,
+            nameExtension: data.nameExtension,
+            access: data.access
+          },
+          {
+            sourceWorkspaceEntityId: data.id
+          }), 'callback')().then(
+            data.success && data.success("WORKSPACE-CREATE")
+          ));
+
+      data.success && data.success("WORKSPACE-CREATE")
+
+      if (data.students.length > 0) {
+        let groupIdentifiers: number[] = [];
+        let studentIdentifiers: string[] = [];
+
+        data.students.map(student => {
+          if (student.type === "student-group") {
+            groupIdentifiers.push(student.id as number);
+          }
+          if (student.type === "student") {
+            studentIdentifiers.push(student.id as string);
+          }
+        });
+
+
+        await promisify(mApi().organizationmanagement.workspaces.students
+          .create(workspace.id, {
+            studentIdentifiers: studentIdentifiers,
+            studentGroupIds: groupIdentifiers
+          }
+          ), 'callback')().then(
+            data.success && data.success("ADD-STUDENTS")
+          );
+        data.success && data.success("ADD-STUDENTS");
```

Comment:
again, data.succcess should only be called once, once it's done, othewise create a new function for progress if you need to keep track of progress